### PR TITLE
Topic/refactoring

### PR DESCRIPTION
**Pull request by Asakiz**

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -126,104 +127,99 @@ type Log struct {
...
-		return nil, err
-	}
-
-	return &probe, nil
+	return processPost(string("/probe"), probe, req)
 }
 
 // GetInfo get updatehub agent general information
-func (c *Client) GetInfo() (*AgentInfo, error) {
+func (c *Client) GetInfo() []byte {
```
Comment (otavio) [inline api.go]: Here, you cannot change the interface.

File: api.go [hunk 2]
```diff
@@ -234,3 +240,9 @@ func buildURL(path string) string {
 
 	return fmt.Sprintf("http://%s%s", u, path)
 }
+
+func checkError(err error) {
+	if err != nil {
+		log.Fatal(err)
+	}
+}
```
Comment (otavio) [inline api.go]: @Asakiz I am not against the logging but we need to return on the API if it succeed or error out.

File: api.go [hunk 3]
```diff
@@ -213,19 +220,58 @@ func (c *Client) LocalInstall(filePath string) (*APIState, error) {
...
+		_, _, errs := gorequest.New().Post(buildURL(url)).Send(req).EndStruct(&responseStruct)
+		if len(errs) > 0 {
+			return nil, errs[0]
+		}
+	}
+
+	response, err := json.Marshal(responseStruct)
+	if err != nil {
+		return nil, err
+	}
```
Comment (otavio) [inline api.go]: Why marshal it here and not unmarshal? We could drop Unmarshall on all methods if we move it here, no?

Comment (Asakiz) [reply api.go]: The problem is the return type of the function. When we Unmarshal in the specific function we know exactly the type should be like.

Comment (otavio) [reply api.go]: Well, for this function you can use the `interface {}` no?

File: api.go [hunk 4]
```diff
@@ -234,3 +257,17 @@ func buildURL(path string) string {
...
+
+func decode(output Interface) *mapstructure.Decoder {
+	cfg := &mapstructure.DecoderConfig{
+		Metadata: nil,
+		Result:   &output,
+		TagName:  "json",
+	}
+	decoder, err := mapstructure.NewDecoder(cfg)
+	if err != nil {
+		panic(err)
```
Comment (otavio) [inline api.go]: We cannot panic in any case. We need to report to user it is wrong.

**Review by otavio — APPROVED** [review]

**Pull request merged**